test/performance/fast_function_approximations.cpp

```diff
@@ -0,0 +1,282 @@
+
+        // Reference function
+        Func ref_func{ftt.name + "_ref"};
+        ref_func(x, y) = sum(ftt.make_reference(arg_x, arg_y, arg_z));
+        schedule(ref_func);
+        ref_func.compile_jit();
+        double pipeline_time_ref = benchmark([&]() { ref_func.realize(buffer_out); buffer_out.device_sync(); }, bcfg);
+
+        // Print results for this function
+        printf("      %s           : %9.5f ns per evaluation  [per invokation: %6.3f ms]\n",
+               ftt.name.c_str(),
+               pipeline_time_ref * pipeline_time_to_ns_per_evaluation,
+               pipeline_time_ref * 1e3);
+
+        for (PrecisionToTest &precision : precisions_to_test) {
+            printf(" fast_%s (%8s):", ftt.name.c_str(), precision.name);
+
+            Func approx_func{ftt.name + "_approx"};
+            approx_func(x, y) = sum(ftt.make_approximation(arg_x, arg_y, arg_z, precision.precision));
+            schedule(approx_func);
+            approx_func.compile_jit();
+            // clang-format off
+            double approx_pipeline_time = benchmark([&]() {
+                approx_func.realize(buffer_out);
+                buffer_out.device_sync();
+            }, bcfg);
+            // clang-format on
+
+            // Print results for this approximation.
+            printf(" %9.5f ns per evaluation  (per invokation: %6.3f ms)",
+                   approx_pipeline_time * pipeline_time_to_ns_per_evaluation,
+                   approx_pipeline_time * 1e3);
+
+            // Check for speedup
+            bool should_be_faster = true;
+            if (precision.precision.force_halide_polynomial != 0) {
+                for (Target::Feature f : ftt.force_poly_not_faster_on) {
+                    if (target.has_feature(f)) {
+                        should_be_faster = false;
+                    }
+                }
+            }
+            if (should_be_faster) num_tests++;
+
+            int goodness = 0;
+
+            if (pipeline_time_ref < approx_pipeline_time * 0.90) {
+                printf("   %6.1f%% slower", -100.0f * (1.0f - approx_pipeline_time / pipeline_time_ref));
+                if (!should_be_faster) {
+                    printf("  (expected)");
+                    goodness = 1;
+                } else {
+                    printf("!!");
+                    goodness = 0;
+                }
+            } else if (pipeline_time_ref < approx_pipeline_time * 1.10) {
+                printf("   equally fast (%+5.1f%% faster)",
+                       100.0f * (1.0f - approx_pipeline_time / pipeline_time_ref));
+                if (should_be_faster) num_passed++;
+                goodness = 1;
+            } else {
+                printf("   %4.1f%% faster",
+                       100.0f * (1.0f - approx_pipeline_time / pipeline_time_ref));
+                if (should_be_faster) num_passed++;
+                goodness = 2;
+            }
+
+            switch (goodness) {
```

Comment:
Why is this a separate switch statement as opposed to just putting the prints in the if-else above?